File: problems/coding/017-trapping-rain-water/solution.js
function trap(h) {
  const n = h.length; const l = Array(n).fill(0), r = Array(n).fill(0);
  for (let i = 1; i < n; i++) l[i] = Math.max(l[i-1], h[i-1]);
  for (let i = n-2; i >= 0; i--) r[i] = Math.max(r[i+1], h[i+1]);
  let best = 0;
  for (let i = 0; i < n; i++) best += Math.max(0, Math.min(l[i], r[i]) - h[i]);
  return best;
}

File: problems/coding/017-trapping-rain-water/solution.py
def trap(h):
    n = len(h); l = [0]*n; r = [0]*n; best = 0
    for i in range(1, n): l[i] = max(l[i-1], h[i-1])
    for i in range(n-2, -1, -1): r[i] = max(r[i+1], h[i+1])
    for i in range(n): best += max(0, min(l[i], r[i]) - h[i])
    return best
